File: test/e2e/ts/test_basic.ts
/*
 * Phase 1 acceptance test for edvabe — TypeScript mirror of
 * test/e2e/python/test_basic.py. Exercises the same hot path:
 *
 *     create sandbox → run commands → read/write files → list → PTY → watch → kill
 *
 * Expected environment (set by `make test-e2e-ts` or manually):
 *
 *     E2B_API_URL=http://localhost:3000
 *     E2B_DOMAIN=localhost:3000
 *     E2B_API_KEY=edvabe_local
 *     E2B_SANDBOX_URL=http://localhost:3000
 *
 * The SANDBOX_URL override makes the SDK route envd data-plane calls
 * through edvabe on plain HTTP instead of the default
 * `https://49983-<id>.<domain>` host form. The SDK still sends the
 * `E2b-Sandbox-Id` / `E2b-Sandbox-Port` headers edvabe dispatches on.
 */

import { test } from 'node:test'
import assert from 'node:assert/strict'
import { setTimeout as sleep } from 'node:timers/promises'

import { Sandbox } from 'e2b'

async function withSandbox<T>(fn: (sbx: Sandbox) => Promise<T>): Promise<T> {
  const sbx = await Sandbox.create({ timeoutMs: 60_000 })
  try {
    return await fn(sbx)
  } finally {
    try {
      await sbx.kill()
    } catch {
      /* ignore */
    }
  }
}

test('create and kill', async () => {
  const sbx = await Sandbox.create({ timeoutMs: 60_000 })
  assert.ok(sbx.sandboxId, 'sandboxId should be set')
  await sbx.kill()
})

test('commands.run echo', async () => {
  await withSandbox(async (sbx) => {
    const result = await sbx.commands.run('echo hello from edvabe')
    assert.equal(result.stdout.trim(), 'hello from edvabe')
    assert.equal(result.exitCode, 0)
  })
})

test('files.write and read', async () => {
  await withSandbox(async (sbx) => {
    await sbx.files.write('/home/user/foo.txt', 'hello')
    const contents = await sbx.files.read('/home/user/foo.txt')
    assert.equal(contents, 'hello')
  })
})

test('files.list', async () => {
  await withSandbox(async (sbx) => {
    await sbx.files.write('/home/user/foo.txt', 'hello')
    const entries = await sbx.files.list('/home/user')
    assert.ok(
      entries.some((e) => e.name === 'foo.txt'),
      'foo.txt not in listing'
    )
  })
})

test('pty create + send_stdin', async () => {
  await withSandbox(async (sbx) => {
    const chunks: Uint8Array[] = []
    const handle = await sbx.pty.create({
      rows: 24,
      cols: 80,
      onData: (data) => {
        chunks.push(data)
      },
      timeoutMs: 15_000,
    })
    try {
      await sbx.pty.sendInput(handle.pid, new TextEncoder().encode('echo in pty\n'))
      await sbx.pty.sendInput(handle.pid, new TextEncoder().encode('exit\n'))

      const deadline = Date.now() + 10_000
      while (Date.now() < deadline) {
        const joined = Buffer.concat(chunks.map((c) => Buffer.from(c))).toString('utf8')
        if (joined.includes('in pty')) {
          return
        }
        await sleep(100)
      }
      const joined = Buffer.concat(chunks.map((c) => Buffer.from(c))).toString('utf8')
      assert.fail(`pty output did not include 'in pty': ${JSON.stringify(joined)}`)
    } finally {
      try {
        await handle.kill()
      } catch {
        /* ignore */
      }
    }
  })
})

test('files.watchDir', async () => {
  await withSandbox(async (sbx) => {
    let seen = false
    const watcher = await sbx.files.watchDir('/home/user', (event) => {
      if (event.name === 'bar.txt') {
        seen = true
      }
    })
    try {
      await sbx.files.write('/home/user/bar.txt', 'x')
      const deadline = Date.now() + 10_000
      while (Date.now() < deadline && !seen) {
        await sleep(100)
      }
      assert.ok(seen, "did not receive filesystem event for bar.txt")
    } finally {
      try {
        await watcher.stop()
      } catch {
        /* ignore */
      }
    }
  })
})

File: test/e2e/python/test_basic.py
"""
Phase 1 acceptance test for edvabe — exercises the same hot path the
docs/06-phases.md acceptance criterion calls out:

    create sandbox → run commands → read/write files → list → PTY → watch → kill

Expected environment (set by `make test-e2e-python` or manually):

    E2B_API_URL=http://localhost:3000
    E2B_DOMAIN=localhost:3000
    E2B_API_KEY=edvabe_local
    E2B_SANDBOX_URL=http://localhost:3000

The SANDBOX_URL override makes the SDK route envd data-plane calls
through edvabe on plain HTTP instead of the default
`https://49983-<id>.<domain>` form. The SDK still sends the
`E2b-Sandbox-Id` / `E2b-Sandbox-Port` headers edvabe dispatches on.

Each feature lives in its own test so a failure points at the specific
layer to fix (proxy, envd-in-Docker behavior, control plane handler).
"""

import threading
import time

import pytest
from e2b import Sandbox
from e2b.sandbox.commands.command_handle import PtySize


@pytest.fixture
def sbx():
    s = Sandbox.create(timeout=60)
    try:
        yield s
    finally:
        try:
            s.kill()
        except Exception:
            pass


def test_create_and_kill():
    s = Sandbox.create(timeout=60)
    assert s.sandbox_id
    s.kill()


def test_commands_run_echo(sbx):
    result = sbx.commands.run("echo hello from edvabe")
    assert result.stdout.strip() == "hello from edvabe"
    assert result.exit_code == 0


def test_files_write_read(sbx):
    sbx.files.write("/home/user/foo.txt", "hello")
    assert sbx.files.read("/home/user/foo.txt") == "hello"


def test_files_list(sbx):
    sbx.files.write("/home/user/foo.txt", "hello")
    entries = sbx.files.list("/home/user")
    assert any(e.name == "foo.txt" for e in entries)


def test_pty(sbx):
    chunks: list[bytes] = []
    handle = sbx.pty.create(size=PtySize(rows=24, cols=80), timeout=15)

    def drain():
        try:
            handle.wait(on_pty=lambda data: chunks.append(data))
        except Exception:
            pass

    t = threading.Thread(target=drain, daemon=True)
    t.start()
    try:
        sbx.pty.send_stdin(handle.pid, b"echo in pty\n")
        sbx.pty.send_stdin(handle.pid, b"exit\n")
        t.join(timeout=10)
        output = b"".join(chunks)
        assert b"in pty" in output, f"pty output did not include 'in pty': {output!r}"
    finally:
        try:
            handle.kill()
        except Exception:
            pass


def test_watch_dir(sbx):
    watcher = sbx.files.watch_dir("/home/user")
    try:
        sbx.files.write("/home/user/bar.txt", "x")
        deadline = time.monotonic() + 10
        seen = False
        while time.monotonic() < deadline and not seen:
            for evt in watcher.get_new_events():
                if evt.name == "bar.txt":
                    seen = True
                    break
            if not seen:
                time.sleep(0.2)
        assert seen, "did not receive filesystem event for bar.txt"
    finally:
        try:
            watcher.stop()
        except Exception:
            pass
